lib/Dialect/Calyx/Export/CalyxEmitter.cpp

```diff
@@ -585,12 +596,87 @@ void Emitter::emitComponentPorts(ComponentOp op) {
   emitPorts(op.getOutputPortInfo());
 }
 
+/// Emit a primitive extern
+void Emitter::emitPrimitiveExtern(hw::HWModuleExternOp op) {
+  Attribute filename = op->getAttrDictionary().get("filename");
+  indent() << "extern " << filename << space() << LBraceEndL();
+  addIndent();
+  indent() << "primitive " << op.getName();
+
+  if (!op.getParameters().empty()) {
+    os << "[";
+    llvm::interleaveComma(op.getParameters(), os,
+        [&](Attribute param) {
+          auto paramAttr = param.cast<hw::ParamDeclAttr>();
+          os << paramAttr.getName().str();
+        }
+        );
+    os << "]";
+  }
+  os << getAttributes(op);
+  // Emit the ports.
+  emitPrimitivePorts(op);
+  os << ";\n";
+  reduceIndent();
+  os << RBraceEndL();
+}
+
+/// Emit the ports of a component.
+void Emitter::emitPrimitivePorts(hw::HWModuleExternOp op) {
+  auto emitPorts = [&](auto ports) {
+    os << LParen();
+    for (size_t i = 0, e = ports.size(); i < e; ++i) {
+      const hw::PortInfo &port = ports[i];
+
+      os << getAttributes(op) << port.name.getValue()
+         << colon();
+      // We only care about the bit width in the emitted .futil file.
+      // Emit parameterized or non-parameterized bit width.
+      if (hw::isParametricType(port.type)) {
+        auto bitWidth = port.type.cast<hw::IntType>().getWidth().dyn_cast<hw::ParamDeclRefAttr>();
+        os << bitWidth.getName().str();
+      } else {
+        auto bitWidth = port.type.getIntOrFloatBitWidth();
+        os << bitWidth;
+      }
+
+      if (i + 1 < e)
+        os << comma();
+    }
+    os << RParen();
+  };
+  emitPorts(op.getPorts().inputs);
+  os << arrow();
+  emitPorts(op.getPorts().outputs);
+}
+
 void Emitter::emitInstance(InstanceOp op) {
   indent() << getAttributes(op) << op.instanceName() << space() << equals()
            << space() << op.componentName() << LParen() << RParen()
            << semicolonEndL();
 }
 
+void Emitter::emitPrimitive(PrimitiveOp op) {
+  indent() << getAttributes(op) << op.instanceName() << space() << equals()
+           << space() << op.primitiveName() << LParen();
+
+  if (op.parameters().hasValue()) {
+    llvm::interleaveComma(op.parameters().getValue(), os,
+        [&](Attribute param) {
+          auto paramAttr = param.cast<hw::ParamDeclAttr>();
+          auto value = paramAttr.getValue();
+          if (auto intAttr = value.dyn_cast<IntegerAttr>()) {
+            os << intAttr.getInt();
+          } else if (auto fpAttr = value.dyn_cast<FloatAttr>()) {
+            os << fpAttr.getValue().convertToFloat();
+          }
```

Comment:
Perhaps llvm_unreachable here instead of doing nothing?

https://llvm.org/docs/CodingStandards.html#assert-liberally